### PR TITLE
Fix migration tasks

### PR DESCRIPTION
**Pull request by bogdanvlviv**
1) Fix `bin/rails db:forward` first migration.

    When `rails db:rollback` all migrations and then run
    `bin/rails db:forward`
    it doesn't work because we search current_migration by version.
    Migration with 0 version doesn't exist, because of that we can't
    to know current_migration index.
    When current_version migration is 0, we should start from 0.

2) Add additional `raise UnknownMigrationVersionError`

    Raise error on the movement of migrations
    when the current migration does not exist.

    This situation happens often in development when to use `git checkout branch_name` :)
    After deleted last migration's file (with git for example)
    `bin/rails db:rollback` is confusing. It doesn't logs anything and effect on `db/schema.rb` by change version of current migration.

3) Show description `rake db:forward` on `rake -T`

    I decided to show the description because `rake db:forward` is useful with `rake db:rollback`.
    Currently `rake -T` is showing the description for `rake db:rollback`. Should be consistency.

### CODE REVIEW

File: activerecord/lib/active_record/railties/databases.rake [hunk 1]
```diff
@@ -145,7 +145,7 @@ db_namespace = namespace :db do
     db_namespace["_dump"].invoke
   end
 
-  # desc 'Pushes the schema to the next version (specify steps w/ STEP=n).'
+  desc "Pushes the schema to the next version (specify steps w/ STEP=n)."
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/railties/databases.rake]: We explicitly removed the description of this rake task and we still have the same reason to do so. Please keep it hidden. It is not common enough. Users usually use `db:migrate` not `db:forward`

**Comment by rafaelfranca** [comment]
I'm not sure if 2 is useful. If the migration was deleted from the codebase what users can do after seem the exception?

And 3, it is no go. We don't want to promote `db:forward`. `db:migrate` is what they should be using.

**Review by rafaelfranca — CHANGES_REQUESTED** [review]

**Comment by bogdanvlviv** [comment]
@rafaelfranca thanks for the review. I just removed 3.


2 Let's investigate.

Example:
`$ bin/rails generate model user username:string password:string`
`$ bin/rails db:migrate`
```bash
== 20170225222543 CreateUsers: migrating ======================================
-- create_table(:users)
   -> 0.0075s
== 20170225222543 CreateUsers: migrated (0.0078s) =============================
```
db/schema.rb
```ruby
# ...
ActiveRecord::Schema.define(version: 20170225222543) do
# ...
```
Commit code,
then
`$ git branch new_branch`
`$ git checkout new_branch`

`$ bin/rails generate migration add_email_to_users email:string`
`$ rm db/migrate/20170225221040_add_email_to_users.rb`


```bash
== 20170225223125 AddEmailToUsers: migrating ==================================
-- add_column(:users, :email, :string)
   -> 0.0010s
== 20170225223125 AddEmailToUsers: migrated (0.0073s) =========================
```
Commit this!
then
`$ git checkout previous_branch`

## Actual behavior

`$ rake db:rollback` # there is nothing, but

`$ git diff`
```diff
--- a/db/schema.rb
+++ b/db/schema.rb

-ActiveRecord::Schema.define(version: 20170225222543) do
+ActiveRecord::Schema.define(version: 20170225223125) do

   create_table "users", force: :cascade do |t|
     t.string "username"
     t.string "password"
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
+    t.string "email"
   end

 end
```

When we run `$ rake db:rollback` in this situation is rollback nothing , logs nothing, only db/schema.rb is changed. i believe that this behavior is incorrect and confusing a lot. We should raise error.

## Implemented behavior

`$ rake db:rollback`
```bash
rake aborted!
ActiveRecord::UnknownMigrationVersionError:

No migration with version number 20170225223125.

/home/bogdanvlviv/active_projects/rails/activerecord/lib/active_record/migration.rb:1092:in `move'
/home/bogdanvlviv/active_projects/rails/activerecord/lib/active_record/migration.rb:998:in `rollback'
/home/bogdanvlviv/active_projects/rails/activerecord/lib/active_record/railties/databases.rake:144:in `block (2 levels) in <top (required)>'
/home/bogdanvlviv/.rvm/gems/ruby-2.4.0@global/gems/rake-12.0.0/exe/rake:27:in `<top (required)>'
/home/bogdanvlviv/.rvm/gems/ruby-2.4.0/bin/ruby_executable_hooks:15:in `eval'
/home/bogdanvlviv/.rvm/gems/ruby-2.4.0/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => db:rollback
(See full trace by running task with --trace)
```

`$ git diff` # there is nothing

**Comment by bogdanvlviv** [comment]
@rafaelfranca could you please review fix?

**Comment by rafaelfranca** [comment]
Thank you for the explanation. When the users get that error what should they do?

**Comment by bogdanvlviv** [comment]
Users will know about problem faster. They will already investigate where and why migration was lost instead of investigating which kind of problem they have.

**Comment by rafaelfranca** [comment]
But what is the problem? If the code is not in the master branch yet there is no problem yet. They now know there is a problem but what they will have to do if the migration is not supposed to be there?

**Comment by bogdanvlviv** [comment]
They will go to the branch where the migration exists and will be available to do "move of migration" this will help to save the development database valid.

**Comment by rafaelfranca** [comment]
But you are assuming this branch exists, sometimes this branch doesn't exist and the user really meant to delete an existing migration. Now they are going to receive an exception and there is nothing they can do to remove that exception. Unless I'm missing something this exception is not helpful at all.

**Comment by bogdanvlviv** [comment]
Currently `rails db:rollback` doesn't work when current migration is missing. it  only change `db/schema.rb` because the database isn't valid. If not raise error, users will have same problems but don't know it without investigating.

**Comment by rafaelfranca** [comment]
I'm failing to understand why users will have a problem if the `db/schema.rb` changes when they explicitly deleted the migration.

But my question remains. When users see that exception what should they do? Right now the exception give no idea of what should be done. Changing branches to find why the exception was raised is not something they can do because maybe they get this exception without changing branches.

I'm against of adding an exception that is impossible to avoid without clear instructions to the user.

**Comment by bogdanvlviv** [comment]
> When users see that exception what should they do?

It is personally we shouldn't care about it!

When we raise error we just help to understand what is happening. Raise error, in this case, does not affect existing behavior, this `raise` just "notification" that helps to know what is wrong without investigating. This will save time when this situation happen.

I would like to add this `raise` because. would be better if Rails will say something in this case, instead do nothing as it is currently.

> Currently we can't to do `rails db:rollback` or `rails db:forward` if current migration doesn't exist. This is confusign and adding notification as `raise` will helpful to understand what is happen.


@rafaelfranca please take a look at that again. i will remove "2" if you against it.

**Comment by bogdanvlviv** [comment]
Would love to finish this Pull Request. Could you please provide feedback if you have a time?

**Comment by rafaelfranca** [comment]
👍 Let's give it a try in 5.2. Could you rebase your PR?

**Comment by bogdanvlviv** [comment]
@rafaelfranca thank you for the response! Branch rebased.

File: activerecord/lib/active_record/migration.rb [hunk 2]
```diff
@@ -1085,13 +1085,21 @@ def migrations(paths)
 
       def move(direction, migrations_paths, steps)
         migrator = new(direction, migrations(migrations_paths))
-        start_index = migrator.migrations.index(migrator.current_migration)
 
-        if start_index
-          finish = migrator.migrations[start_index + steps]
-          version = finish ? finish.version : 0
-          send(direction, migrations_paths, version)
+        if current_version != 0 && migrator.current_migration.nil?
```
Comment (bogdanvlviv) [inline activerecord/lib/active_record/migration.rb]: Good point. Fixed!

**Review by rafaelfranca — APPROVED** [review]

**Pull request merged**